Level 3

skan7575 opened the pull request.

Commit:
Пока все норм

Commit:
Пока все норм

Commit:
Пока все норм

Commit:
Пока все норм

Commit:
Пока все норм

Commit:
Пока все норм

Commit:
Готово!

Commit:
Готово!

Commit:
Исправлен Readme

SergeyK2004 commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 

    **Комментарий: при каждом поиске на странице фильмов результаты поиска фильмов нужно сохранять в localstorage. В сохраняемые данные должны входить и строка поиска и статус чекбокса «Короткометражки», чтобы пользователю при повторном переходе на страницу фильмов, можно было вывести на экран информацию, полностью совпадающую с его последним поиском (с расстановкой актуального состояния лайков) перед уходом на другую страницу или перезагрузкой страницы. Эту информацию можно будет доставать из хранилища при монтировании и перерисовке компоненты фильмов. При логауте пользователя нужно эту информацию в хранилище и стейте (useState, если она там есть) уничтожать.**

    **Сейчас пользователь не видит точного результата своего последнего поиска на странице фильмов после перезагрузки страницы фильмов. Переключатель "Короткометражка" может быть переключен в "Сохраненных" и не восстанавливается при возвращении на "Фильмы". Найденные фильмы не созраняются в хранилище.**

	**Замечание: Сохранять данные последнего поиска на странице «Сохранённые фильмы» для их показа при повторном входе на эту страницу не нужно. Надо при входе на эту страницу выводить на экран все сохранённые фильмы, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на страницу «Сохранённые фильмы» пользователь видел все свои сохранённые карточки без задания параметров поиска.**

	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
**Комментарий: не корректно работает валидация формы поиска, по заданию "После сабмита формы поиска производится валидация. Если в поле не введён текст, выводится ошибка «Нужно ввести ключевое слово»." https://skr.sh/sH17nKNcjON Вместо этого включена браузерная валидация. В вашем проекте браузерной валидации не должно быть. Все ошибки должны обрабатываться и показываться только средствами приложения**
   - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
    **Комментарий: https://skr.sh/sGTAOX7Gk4k Согласно вашему заданию ошибки валидации должны указываться под полем, других вариантов не предусмотрено. https://skr.sh/sH1Fjh5yy83 Кнопка сохранения не должна быть активной, пока все поля не будут валидны**
   - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    **Комментарий: При открытии форма должна заполняться данными пользователя, записывать их в плэйсхолдер не нужно.**
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	**Комментарий: при сохранение профиля пользователь должен увидеть сообщение о подтверждении, вид сообщения может быть любым на ваш выбор**
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
	**Комментарий: https://skr.sh/sH1XSrbh1EX кнопка активна даже при пустых полях ввода**
   - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	**Комментарий: при каждом поиске фильмов данные запрашиваются с сервера, не нужно повторно делать запрос, нужно использовать данные предыдущего т.к. данные сервер возвращает те же самые - по заданию "При запросе к серверу за фильмами на странице «Фильмы» вы получаете сразу все данные и сохраняете их."**
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
	**Комментарий: по заданию количество карточек должно зависеть от размера экрана, сейчас может добавляться 3 картчоки на разрешении, при котором в ряду отображается толкьо две.**
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
  **Комментарий: На странице "Фильмы" лайки могут быть показы оставшиеся от предыдущего пользователя. https://skr.sh/sH1nirLeoAb У текущего нового пользователя нет ни одного сохраненного фильма, однако лайки на фильмах стоят в соответствии с сохраненными фильмами прошлого пользователя, который уже выполнил логаут.**
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
    **Комментарий: Если при постановке лайка получена ошибка 401 необходимо выполнить действия по логауту пользователя с удалением данных из хранилища и очисткой стэйтов. Сейчас при некорректном токине и получении ошибок о нарушении авторизации пользователь может продолжать "лайкать" фильмы. При любых других ошибках выполнять логаут НЕ нужно. Только в случае получения ошибки авторизации.**
    **Если нет связи с сервером, все равно можно отправлять фильмы в сохраненное и на них будут проставляться лайки**
 - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
	**Комментарий: происходит открытие системной 404 страницы https://skr.sh/sH1kV2SubWB**
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
    **Комментарий: При попытке авторизованного пользователя войти на любую страницу по её URL-адресу в адресной строке, происходит открытие системной 404 страницы https://skr.sh/sH1kV2SubWB**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используются хуки роутинга.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
	**Комментарий: кастомная 404 страница не реализована**
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  **Комментарий: Если при любых запросах с проверкой Jwt, сервер приложения возвращает статус ошибки (токен может оказаться невалидным, даже будучи в наличии на клиенте, или может быть уничтожен в localstorage авторизованным пользователем "вручную"), пользователь должен становиться неавторизованным - должен происходить полный "логаут" пользователя на главную страницу с шапкой неавторизованного пользователя с удалением всей его информации из хранилища и стейтов. Сейчас в случае невалидного или уничтоженного токена в localstorage, после перезагрузки сайта, логаут происходит не полностью https://skr.sh/sH1yHpK0dxM**
  **Комментарий: проверки, что jwt есть в хранилище, недостаточно, при неправильном токене не происходит редирект на страницу входа**
    - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
  **Комментарий: При каждом поисковом запросе выполняется обращение к beatfilm-movies**
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [ ] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
**Комментарий: на страницах с фильмами не отображается активная страница, используйте cтили NavLink для задания активности**
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [ ] <!--4--> пользователь получает сообщение в случае любой ошибки;
    **Комментарий: если авторизация, регистрация или редактирование профиля завершились ошибкой, пользователь должен увидеть сообщение**
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
	**Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
    **Комментарий: не нужно очищать поля формы до ответа сервера подтверждением, у пользователя должна быть возможность исправить данные, а не вводить их заново. На время выполнения запросанужно только блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при поиске или нажатии кнопки "Еще" на разных размерах экрана и точки перестроения экрана нужно вынести в константы**

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 44

Commit:
Исправил сервер, исправил валидацию профиля и кнопку, добавил валидацию input Search

Commit:
Исправил сервер исправил валидацию профиля и кнопку, добавил валидацию input Search исправил валидацию токена при обновлении страницы а также запросах.

Commit:
Исправил сервер
исправил валидацию профиля и кнопку,
добавил валидацию input Search
исправил валидацию токена при обновлении страницы а также запросах.
Исправил валидацию форм, у меня кнопка всегда серая при пустых полях там стоит checkvaliditu
попрафил форму обновления профиля

Commit:
Исправил сервер
исправил валидацию профиля и кнопку,
добавил валидацию input Search
исправил валидацию токена при обновлении страницы а также запросах.
Исправил валидацию форм, у меня кнопка всегда серая при пустых полях там стоит checkvaliditu
попрафил форму обновления профиля
Исправил все основные ошибки

Commit:
прошел eslint

Commit:
Исправил ошибку 401

Commit:
Исправил последнее успел найти

Commit:
исправил ничего не найдно

SergeyK2004 commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
   - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
   - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
   - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
 - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используются хуки роутинга.
  - [ ] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
	**Комментарий: 404 страница отображается, но сразу происходит перезход на главную, если пользователь неавторизован.**
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
    - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
	**Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
    **Комментарий: не нужно очищать поля формы до ответа сервера подтверждением, у пользователя должна быть возможность исправить данные, а не вводить их заново. На время выполнения запросанужно только блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при поиске или нажатии кнопки "Еще" на разных размерах экрана и точки перестроения экрана нужно вынести в константы**

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 94

## Отлично, теперь необходимый функционал работает. Поздравляю с завершением дипломного проекта!


==========================================================

Однако, если планируете использовать проект в портфолио, то советую ещё поработать над ним, исправить оставшиеся замечания и ещё раз перепроверить функционал

=================================================================


### Несколько советов для дальнейшего развития:
- сверстайте несколько небольших проектов для себя, для закрепления знаний и что бы можно добавить их в портфолио
- еще немного попрактиковаться с React, например изучить redux
- будет большим плюсом при поиске работы, если немного изучите TypeScript т.к. на него переходят многие компании
- ходить на собеседования, опыт прохождения собеседований не менее важен, чем знание теории, и Вы поймете какие знания требуются на рынке и где Вы проседаете
- читайте [habr](https://habr.com/ru/) и слушайте подкаксты, они очень хорошо расширяют кругозор, и позволяют узнать, что происходит в отрасли:
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

skan7575 commented:
-

skan7575 committed:
Merge branch 'main' into level-3
